pkg/acsengine/pki.go

```diff
@@ -52,61 +52,119 @@ func CreatePki(extraFQDNs []string, extraIPs []net.IP, clusterDomain string, caP
 		clientPrivateKey      *rsa.PrivateKey
 		kubeConfigCertificate *x509.Certificate
 		kubeConfigPrivateKey  *rsa.PrivateKey
+		etcdServerCertificate *x509.Certificate
+		etcdServerPrivateKey  *rsa.PrivateKey
+		etcdClientCertificate *x509.Certificate
+		etcdClientPrivateKey  *rsa.PrivateKey
+		etcdPeerCertPairs     []*PkiKeyCertPair
 	)
 	errors := make(chan error)
 
 	var err error
 	caCertificate, err = pemToCertificate(caPair.CertificatePem)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, nil, nil, nil, nil, nil, err
 	}
 	caPrivateKey, err = pemToKey(caPair.PrivateKeyPem)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, nil, nil, nil, nil, nil, err
 	}
 
 	go func() {
 		var err error
-		apiServerCertificate, apiServerPrivateKey, err = createCertificate("apiserver", caCertificate, caPrivateKey, true, extraFQDNs, extraIPs, nil)
+		apiServerCertificate, apiServerPrivateKey, err = createCertificate("apiserver", caCertificate, caPrivateKey, false, true, extraFQDNs, extraIPs, nil)
 		errors <- err
 	}()
 
 	go func() {
 		var err error
 		organization := make([]string, 1)
 		organization[0] = "system:masters"
-		clientCertificate, clientPrivateKey, err = createCertificate("client", caCertificate, caPrivateKey, false, nil, nil, organization)
+		clientCertificate, clientPrivateKey, err = createCertificate("client", caCertificate, caPrivateKey, false, false, nil, nil, organization)
 		errors <- err
 	}()
 
 	go func() {
 		var err error
 		organization := make([]string, 1)
 		organization[0] = "system:masters"
-		kubeConfigCertificate, kubeConfigPrivateKey, err = createCertificate("client", caCertificate, caPrivateKey, false, nil, nil, organization)
+		kubeConfigCertificate, kubeConfigPrivateKey, err = createCertificate("clientclient", caCertificate, caPrivateKey, false, false, nil, nil, organization)
 		errors <- err
 	}()
 
+	go func() {
+		var err error
+		organization := make([]string, 1)
+		organization[0] = "system:masters"
+		ip := net.ParseIP("127.0.0.1").To4()
+		peerIPs := append(extraIPs, ip)
+		etcdServerCertificate, etcdServerPrivateKey, err = createCertificate("etcdserver", caCertificate, caPrivateKey, true, true, nil, peerIPs, organization)
+		errors <- err
+	}()
+
+	go func() {
+		var err error
+		organization := make([]string, 1)
+		organization[0] = "system:masters"
+		ip := net.ParseIP("127.0.0.1").To4()
+		peerIPs := append(extraIPs, ip)
+		etcdClientCertificate, etcdClientPrivateKey, err = createCertificate("etcdclient", caCertificate, caPrivateKey, true, false, nil, peerIPs, organization)
+		errors <- err
+	}()
+
+	etcdPeerCertPairs = make([]*PkiKeyCertPair, masterCount)
+	for i := 0; i < masterCount; i++ {
+		go func(i int) {
+			var err error
+			organization := make([]string, 1)
+			organization[0] = "system:masters"
+			ip := net.ParseIP("127.0.0.1").To4()
```

Comment:
It is a bit strange that the IPs for client, server and peers are all 127.0.0.1. In a deployment environment, do all master nodes use the same 127.0.0.1? It may worth some investigation later for this.

It might also worth (?) to wrap these cert generation logic in another helper function since they are duplicated for 3 times here.

Reply:
I'm confused, what do you mean by "It is a bit strange that the IPs for client, server and peers are all 127.0.0.1"?

127.0.0.1 is the loopback address so every machine has the same one. Here I'm adding it to the list of peer IP addresses since etcd tls does client-server and peer-to-peer authentication.

Reply:
Sorry for the confusion, I though this IP is used for the identity in the cert, it looks it is used in SAN, so it should be good.